etcd/src/main/scala/io/buoyant/etcd/Key.scala

```diff
@@ -0,0 +1,278 @@
+      boolParam(wait).map("wait" -> _),
+      waitIndex.map("waitIndex" -> _.toString)
+    ).flatten
+    val req = mkReq(uriPath, params = params)
+    req.headerMap("accept") = MediaType.Json
+    client(req).flatMap { rsp => Future.const(NodeOp.mk(req, rsp, key, params)) }
+  }
+
+  /**
+   * Set the contents of a key.
+   *
+   * If value is None, the key is treated as a directory.  In order to
+   * unset the value of a data node, use `Some(Buf.Empty)`.
+   *
+   * Optionally, a `ttl` may be specified to inform etcd to remove the
+   * node after some time period (only second-granularity is supported
+   * by etcd).
+   *
+   * If `prevExist` is true, the node operation will fail if the node
+   * does not already exist.
+   */
+  def set(
+    value: Option[Buf],
+    ttl: Option[Duration] = None,
+    prevExist: Boolean = false
+  ): Future[NodeOp] = {
+    var params = valueParams(value)
+    for (ttl <- ttl) {
+      val v = if (ttl == Duration.Zero) "" else ttl.inSeconds.toString
+      params = params :+ "ttl" -> v
+      if (prevExist) {
+        params = params :+ "prevExist" -> "true"
+      }
+    }
+    val req = mkReq(uriPath, Method.Put, params)
+    client(req).flatMap { rsp => Future.const(NodeOp.mk(req, rsp, key, params)) }
+  }
+
+  /**
+   * Create a new key.
+   *
+   * If the key exists, an error is returned.
+   *
+   * If value is None, the key is treated as a directory.  In order to
+   * create an empty data node, use `Some(Buf.Empty)`.
+   *
+   * Optionally, a `ttl` may be specified to inform etcd to remove the
+   * node after some time period (only second-granularity is supported
+   * by etcd).
+   */
+  def create(
+    value: Option[Buf],
+    ttl: Option[Duration] = None
+  ): Future[NodeOp] = {
+    val vp = valueParams(value)
+    val tp = ttl.map("ttl" -> _.inSeconds.toString)
+    val params = vp ++ Seq(tp).flatten
+
+    val req = mkReq(uriPath, Method.Post, params)
+    client(req).flatMap { rsp => Future.const(NodeOp.mk(req, rsp, key, params)) }
+  }
+
+  /**
+   * Set the node's data if the provided preconditions apply to the
+   * existing state of the node.
+   *
+   * If `prevIndex` is specified, the current node must have the
+   * provided `index` value.
+   *
+   * If `prevValue` is specified, the current node must have the
+   * provided value.
+   *
+   * If `prevExist` is specified, the node must already exist.
+   */
+  def compareAndSwap(
+    value: Buf,
+    prevIndex: Option[Long] = None,
+    prevValue: Option[Buf] = None,
+    prevExist: Boolean = true
+  ): Future[NodeOp] = {
+    require(prevIndex.isDefined || prevValue.isDefined || !prevExist)
+
+    var params = Seq.empty[(String, String)]
+    for (i <- prevIndex) {
+      params = params :+ "prevIndex" -> i.toString
+    }
+    for (Buf.Utf8(v) <- prevValue) {
+      params = params :+ "prevValue" -> v
+    }
+    if (!prevExist) {
+      params = params :+ "prevExist" -> "false"
+    }
+    val Buf.Utf8(v) = value
```

Comment:
yeah, this always matches, just sometimes with funny characters.  turns out it's really hard to test whether a byte array contains a utf8 string